### PR TITLE
LIP-25: Governor contract

### PR DESCRIPTION
**Pull request by kyriediculous**
This PR implements the `Governor` contract according to the spec defined in LIP25 https://github.com/livepeer/LIPs/pull/33

### CODE REVIEW

File: contracts/zeppelin/ECTools.sol
Comment (yondonfu) [inline contracts/zeppelin/ECTools.sol]: Looks like we can remove this file.

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: Let's place this directive above the rest of the state variable declarations to be consistent with the style used by the other contracts in this repo. Would be nice to have the Solidity linter check for this...not sure if it supports this style rule though.

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: Let's update this comment because we're now using msg.sender based authentication instead of signature based authentication for the updates.

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: The linter is [complaining](https://app.circleci.com/pipelines/github/livepeer/protocol/20/workflows/a4dd81b5-ddc3-40a1-935b-453c40ee1fcc/jobs/1410/steps) about this line because the security plugin by default disallows the use of `.call.value()`. I actually now disagree with this style rule (given the need to use it in settings such as this one and given that when just sending ETH `.transfer()`'s gas stipend can introduce issues due to opcode gas cost changes) so I'd be in favor of updating `.soliumrc.json` to turn off the rule altogether.

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: Having the nonce in the update is useful to distinguish between updates that could contain the same tx data, but I'm not sure that the contract needs to keep track of a global nonce. Do you have a scenario that you're thinking of that global nonce tracking would be required for?

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: I see that the comment style here is different from the style used in the rest of the file. Can we be consistent with the comment style we use?

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js]: Looks like we can remove all references to ECTools.

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js]: If we're already pulling in `ethers`, it could be clearer to use `ethers.constants.MaxUint256.add(utils.bigNumberify(1)` here.

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js]: "updateCount" seems to imply there is an additional var that is being updated. If that is not the case I suggest removing it.

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: Right, this is not as useful without signature based auth. 

I guess it's useful as an index for the edge case where there's multiple updates in a single block, otherwise blocknumbers would suffice to have historical ordering.

**Comment by kyriediculous** [comment]
moved solium package to ethlint in 93301cc 

https://github.com/duaraghav8/Ethlint

> If you're currently using the solium package for npm install, it is highly recommended that you move to ethlint. The solium package will not receive updates after December, 2019. There are no differences between the updates pushed to ethlint and solium packages.

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js]: fixed in 1de239d

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js]: fixed in cd04b83

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js]: fixed in cd04b83

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: fixed in cd04b83

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: fixed in cd04b83

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: fixed in cd04b83

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: fixed in cd04b83

File: contracts/zeppelin/ECTools.sol
Comment (kyriediculous) [reply contracts/zeppelin/ECTools.sol]: fixed in cd04b83

File: contracts/governance/Governor.sol
Comment (yondonfu) [reply contracts/governance/Governor.sol]: The `UpdateStaged` and `UpdateExecuted` events within a single block should be emitted in the order that updates are staged or executed. I believe this ordering is achieved by attaching an index to each log within a block (would be good to double check this though).

Furthermore, I think having a nonce included in the update makes sense, but I don't think the contract needs to track a global nonce and do any checks on the nonce included in an update.

**Review by yondonfu — CHANGES_REQUESTED** [review]
Looks like Coveralls is complaining about reduced coverage due to uncovered lines in the Governor - see the [latest report](https://coveralls.io/builds/32156721/source?filename=contracts/governance/Governor.sol).

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: Ordering of transactions within a single block are determined by the miner based on transaction nonce and price (and can thus vary between implementations) thus I don't think there's guarantees that log index (or nonce tracked client-side by events) ordering is preserved (Unless `owner` is an EOA but in production this will not be the case, right?). 

The nonce in this contract does provide that guarantee, so I think the question is whether the tradeoff of a `SLOAD` and `STORE` against having hard ordering is worth it. 

The way the delay mapping works in this implementation should provide ample replay and re-entry prevention (it would be good to have this as a highlight for the audit) so it's more a nice to have rather than an explicit security requirement.

**Comment by kyriediculous** [comment]
I will add the necessary tests to bring coverage up.

File: contracts/governance/Governor.sol
Comment (yondonfu) [reply contracts/governance/Governor.sol]: Hm I actually think regular transaction nonces already handle the scenario that you're describing.

2 distinct `stage()`/`update()` calls can be included in the same block in the following scenarios:

1. If the contract owner batched the calls together into a single atomic transaction
2. If the contract owner made two separate calls in two separate transactions and those transactions happen to be included in the same block

In scenario 1, the ordering of the calls is defined by the signer of the transaction.

In scenario 2, regardless of whether the contract owner itself is a contract or an EOA, each of the transactions would have a nonce that guarantees execution order - the transaction that was created/signed first should have a lower nonce and should be executed before the second transaction.

**Review by yondonfu — COMMENTED** [review]
Should there also be a function to cancel staged updates in the scenario that an update is staged, but before it is executed there is a desire to not move forward with it?

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: Let's fix this spacing.

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: In scenario 1 while true, other stakeholders depend on the benevolence of the `owner` to define the right nonces/indexes. There are no barriers of submitting multiple transactions with the same nonce or older nonces if it is not tracked and checked. 

Full transparency does come at a cost, in this case that of a storage read and write.

Scenario 2 does only uphold for EOA's because contract accounts will not increment nonces unless for `CREATE` (and I think `CREATE2` opcodes) 

From the [Ethereum yellow paper](http://gavwood.com/paper.pdf), page 3, section 4.1 'World State' :  

> nonce: A scalar value equal to the number of transactions sent from this address or, in the case
of accounts with associated code, the number of
contract-creations made by this account. For account of address a in state σ, this would be formally denoted σ[a]n.

**Comment by kyriediculous** [comment]
> Should there also be a function to cancel staged updates in the scenario that an update is staged, but before it is executed there is a desire to not move forward with it?

From a feature perspective it's a similar discussion as the nonce in my opinion: 

Is social convention enough for updates cancellation (e.g. staged but never executed upon). 

Or do we want guarantees that a cancelled update can never be executed (explicit cancellation function) 

From a technical perspective having explicit cleanups would be nice for reducing unnecessary state growth. 

But again it comes at a cost. 

Also we'd have to figure out ( and probably put in scope for this LIP) what the cancellation process would look like and how that would work with delays.

File: contracts/governance/Governor.sol
Comment (yondonfu) [reply contracts/governance/Governor.sol]: > There are no barriers of submitting multiple transactions with the same nonce or older nonces if it is not tracked and checked.

Right, but the contract owner can always submit the same update with a nonce that passes the contract nonce check. So, I'm not sure that I see what the contract nonce check protects against in this scenario. Since the contract owner has the authority to call `stage()`, it can stage whatever update that it wants.

> Scenario 2 does only uphold for EOA's because contract accounts will not increment nonces

The scenario I am referring to here involves a contract calling the Governor in two separate transactions (the contract calls against the Governor would be "internal transactions", but those shouldn't be relevant here since we just care about the other transactions that triggered the contract calls in the first place). In order for there to be two separate transactions, an EOA would have to submit two separate transactions to the contract. Each of these transactions will need to have a unique nonce.

**Comment by yondonfu** [comment]
> Or do we want guarantees that a cancelled update can never be executed (explicit cancellation function)

I think this is preferable since after the delay for a staged update any account can call the `execute()` function to  execute the update which means never executing a staged update would be insufficient.

> Also we'd have to figure out ( and probably put in scope for this LIP) what the cancellation process would look like and how that would work with delays.

We could allow a `cancel()` function to be called for a staged update as long as it hasn't been executed yet (in this case the staged update would have been removed from the Governor anyway). One reason why a staged update would be canceled is if during the delay a bug in the implementation was discovered. The social convention would be that the `cancel()` function should only be used for this reason.

**Comment by kyriediculous** [comment]
482bdbb - adds a `cancel(Update memory _update)` function that works similar to the other Governor functions

56c3078 - forwards revert reasons from external calls 

e7c13ef - adds test for our modified `Ownable.sol` implementation (mainly the `onlyThis` modifier) 

5ef3569 - misc fixes

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: fixed in d700e93

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: > Right, but the contract owner can always submit the same update with a nonce that passes the contract nonce check.

That's true but the owner can't submit an update with the same nonce regardless, it's about ordering not security at this point.

The owner could theoretically submit a batch of `stage()` calls within the same transaction using a smart contract wallet. 

I'm okay with removing the nonce notion since it's unlikely this scenario will ever take place though, but let's be mindful of it.

Comment (kyriediculous) [reply contracts/governance/Governor.sol]: removed notion of a nonce in 9050b40

**Comment by kyriediculous** [comment]
This should be ready for final review now.

once this PR is approved we can move it to last call. I'd like to clean up the commit history a little bit after final review though.

File: contracts/governance/Governor.sol
Comment (kyriediculous) [inline contracts/governance/Governor.sol]: Note to self: fix compiler version, this is considered good practice. Not really super important for our use case but if someone compiles this contract with a different compiler version above 0.5.11 installed there _could_ be differences in behaviour (or exposure to vulnerabilities that could be in a specific compiler version that isn't as widely used).

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: Should the revert reason be more specific to differentiate between other authorization related modifiers (i.e. `onlyOwner`)? i.e. "unauthorized: msg.sender not Governor"

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: Should the revert reason be more specific to differentiate between other authorization related modifiers (i.e. `onlyThis`)?

Ex. "unauthorized: msg.sender not owner"

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: "null address" instead of "zero"?

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol:62]: The convention used in this repo is to emit the event last after any effects & interactions - let's continue to follow that convention here?

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol:76]: Since we don't check if the update is already staged, it would be possible for the same update to be staged multiple times which would just set the execute block farther in the future. 

Should we be more strict here and disallow staging an existing update by adding `require(updates[updateHash] == 0, ...)`?

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: This needs an `onlyOwner`.

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js]: Doesn't seem like we need this anymore since updates don't need to be signed.

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js:169]: Let's also check that the contract storage mapping for updates is updated correctly?

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js:77]: Let's also check that the contract storage value for `owner` is updated correctly?

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js]: "emites" -> "emits"

Comment (yondonfu) [inline test/unit/Governor.js]: Let's also check that the contract storage mapping for updates is updated correctly?

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js:537]: Should we also test for successful ETH transfer when the "value" in one of the txs included in an update is non-zero?

File: test/unit/Governor.js
Comment (yondonfu) [inline test/unit/Governor.js:626]: Should we add a test when a remote call other than the first one reverts that checks that the revert reason corresponds with the correct remote call (and not just the first one)?

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol:62]: If that convention requires us to define an intermediate variable `oldOwner` I don't see much purpose in adhering to it since the code in this function is executed atomically anyway.

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: Woops, will add a test for it as well.

File: contracts/governance/Governor.sol
Comment (yondonfu) [reply contracts/governance/Governor.sol:62]: Ah I see. If emitting the event here avoids introducing an additional var then this looks fine to me.

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js:537]: We could but if this was written the wrong way it wouldn't compile, but we could add it for completeness.

Comment (yondonfu) [reply test/unit/Governor.js:537]: I think we should add a test for ETH transfers if the contract is going to support ETH transfers as a part of updates.

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js:626]: I don't see much value here because 

1. execute already tests that `success == true` when the call doesn't fail and thus no revert is encountered
2. we will only ever get the revert reason for the first failing call

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js:169]: slightly annoying due to not being able to use `ethers.utils.solidityKeccak256`  . It only supports packed data but our data type doesn't support packed data. 

The least annoying solution would be to add a pure function to the governor contract (or a seperate one we just deploy for testing) that can calculate the hash for an update given the pre-image.

**Comment by kyriediculous** [comment]
I believe 21161ac addresses all change requests

File: test/unit/Governor.js
Comment (yondonfu) [reply test/unit/Governor.js:77]: Bump

File: test/unit/Governor.js
Comment (yondonfu) [reply test/unit/Governor.js:626]: Fair enough. After some more reflection, I think I'm actually more interested in seeing cases where different the transactions in an update are for different contract and/or different functions - maybe something simple can be included in an integration test.

**Review by yondonfu — COMMENTED** [review]
The implementation is looking solid! 

WDYT about adding a small integration test to simulate a few batch updates for the other Livepeer contracts via the Governor since that will be how it will be used in practice?

Also, the actual design of the Governor looks pretty good to me at this point so I'd be comfortable with moving it to Last Call whenever you would like to do so.

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js:626]: +1 for some integration test scenario's using our actual contracts

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js:77]: fixed in 9c11912

File: test/unit/Governor.js
Comment (kyriediculous) [reply test/unit/Governor.js:626]: I added integration tests but I also added some changes for Testability as mentioned in this comment https://github.com/livepeer/protocol/pull/391#issuecomment-674029279 , I also removed the `whenSystemNotPaused` modifier from `Minter.DepositEth` like in #391. 

These Minter changes for testability are a separate commit and can be dropped if we merge #391 first (and the  revert reasons mentioned are added).

Comment (kyriediculous) [reply test/unit/Governor.js:626]: I also decided to not  test for all `Minter.migrateToNewMinter`  revert paths since that is not the purpose of this test, I added two cases here and I think that's sufficient.

**Comment by kyriediculous** [comment]
e4af94a fixes a typo and some descriptions to match revert reasons

File: test/integration/GovernorUpdate.js
Comment (yondonfu) [inline test/integration/GovernorUpdate.js]: Is this needed?

File: test/integration/GovernorUpdate.js
Comment (yondonfu) [inline test/integration/GovernorUpdate.js:38]: Is this missing a successful case?

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol:52]: WDYT about emitting an `OwnershipTransferred` (from the null address to msg.sender) event here similar to what the OZ Ownable contract [constructor](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol#L29) does? It's just a single line to emit the event and the benefit is that it preserves the invariant that an event is emitted whenever ownership is updated (including in the constructor).

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: test/integration/GovernorUpdate.js
Comment (kyriediculous) [reply test/integration/GovernorUpdate.js]: No my vscode autofill keeps adding this , it's already injected by default.

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol:52]: fixed in cf41a3c

File: test/integration/GovernorUpdate.js
Comment (kyriediculous) [reply test/integration/GovernorUpdate.js:38]: fixed in db643ac

File: test/integration/GovernorUpdate.js
Comment (kyriediculous) [reply test/integration/GovernorUpdate.js]: fixed in db643ac

File: contracts/governance/Governor.sol
Comment (yondonfu) [inline contracts/governance/Governor.sol]: Minor: Think its cheaper to do `emit OwnershipTransferred(address(0), ...)` (as is done in the OZ Ownable constructor) since it avoids reading `owner` and we already know the initial values of `owner` is `address(0)`.

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: contracts/governance/Governor.sol
Comment (kyriediculous) [reply contracts/governance/Governor.sol]: Good point, especially since istanbul.

Fixed in ac98605

**Comment by kyriediculous** [comment]
522ffdb fixes the changes discussed after the security review 

- make `execute` payable instead of `stage`
- introduce a `nonce` to the update struct so that we can differentiate between updates , i.e. unique identifier. Not stored on-chain.

**Comment by kyriediculous** [comment]
> The low-level functions call, delegatecall and staticcall return true as their first return value if the account called is non-existent, as part of the design of the EVM. Account existence must be checked prior to calling if needed.

Should we check the codesize of the target address ?

**Comment by yondonfu** [comment]
> Should we check the codesize of the target address ?

I think its reasonable to just have the update submitter be responsible for ensuring that the contract exists via a check off-chain instead of adding that check in the contract since the update submitter is already responsible for ensuring that the rest of the update (i.e. tx data) is correct.

**Comment by yondonfu** [comment]
Updates look good to me - let's rebase on top of `streamflow` since a few changes have been made there recently?

**Comment by kyriediculous** [comment]
Done !

**Review by yondonfu — APPROVED** [review]

**Pull request merged**
